File: cpp/fibonacci.cpp
#include <iostream>
#include <unordered_map>
#include <chrono>
#include <fstream>
#include <filesystem>  // Para manejar directorios correctamente

// Memoización con unordered_map para mejorar rendimiento
std::unordered_map<int, uint64_t> memo;

uint64_t fibonacci(int n) {
    if (n <= 1) return n;
    if (memo.find(n) != memo.end()) return memo[n];

    memo[n] = fibonacci(n - 1) + fibonacci(n - 2);
    return memo[n];
}

int main() {
    // Medir tiempo de ejecución
    auto start_time = std::chrono::high_resolution_clock::now();

    // Calcular Fibonacci hasta 10,000
    for (int i = 0; i < 10000; i++) {
        fibonacci(i);
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    double execution_time_ms = std::chrono::duration<double, std::milli>(end_time - start_time).count();

    std::cout << execution_time_ms << " ms" << std::endl;

    // Crear directorio de salida usando filesystem (C++17+)
    std::string output_dir = "/app/salida_cpp";
    if (!std::filesystem::exists(output_dir)) {
        if (!std::filesystem::create_directories(output_dir)) {
            std::cerr << "Error: No se pudo crear el directorio de salida." << std::endl;
            return 1;
        }
    }

    // Guardar el tiempo en un archivo
    std::ofstream outfile(output_dir + "/salida_cpp.txt");
    if (outfile.is_open()) {
        outfile << execution_time_ms << " ms"  << std::endl;
        outfile.close();
    } else {
        std::cerr << "Error: No se pudo crear el archivo de salida." << std::endl;
        return 1;
    }

    return 0;
}

File: goland/fibonacci.go
package main

import (
	"fmt"
	"os"
	"time"
)

// Fibonacci recursivo con memoización
func fibonacci(n int, memo map[int]uint64) uint64 {
	if n <= 1 {
		return uint64(n)
	}
	if val, exists := memo[n]; exists {
		return val
	}
	memo[n] = fibonacci(n-1, memo) + fibonacci(n-2, memo)
	return memo[n]
}

func main() {
	startTime := time.Now()

	// Crear un mapa para la memoización
	memo := make(map[int]uint64)

	// Calcular Fibonacci hasta 20,000
	for i := 0; i < 10000; i++ {
		fibonacci(i, memo)
	}

	endTime := time.Since(startTime)
	executionTime := float64(endTime.Nanoseconds()) / 1e6 // Convertir a ms con decimales

	fmt.Printf("%.6f ms\n", executionTime)

	// Crear directorio si no existe
	outputDir := "/app/salida_go"
	if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
		fmt.Println("Error creando directorio:", err)
		return
	}

	// Guardar el tiempo en un archivo
	outputFile := outputDir + "/salida_go.txt"
	file, err := os.Create(outputFile)
	if err != nil {
		fmt.Println("Error creando archivo:", err)
		return
	}
	defer file.Close()
	
	_, err = file.WriteString(fmt.Sprintf("%.6f ms", executionTime))
	if err != nil {
		fmt.Println("Error escribiendo en archivo:", err)
	}
}
